bin/runtime-common/src/refund_relayer_extension.rs

```diff
@@ -241,40 +221,40 @@ where
 		// reject batch transactions with obsolete headers
 		self.validate(who, call, info, len).map(drop)?;
 
-		// now try to check if tx matches one of types we support
-		let parse_call_type = || {
-			if let Some(UtilityCall::<R>::batch_all { ref calls }) = call.is_sub_type() {
-				if calls.len() == 3 {
-					return Some(CallType::AllFinalityAndDelivery(
-						extract_expected_relay_chain_state::<R, GI>(&calls[0])?,
-						extract_expected_parachain_state::<R, GI, PI, PID>(&calls[1])?,
-						extract_messages_state::<R, GI, MI, LID>(&calls[2])?,
-					))
-				}
-				if calls.len() == 2 {
-					return Some(CallType::ParachainFinalityAndDelivery(
-						extract_expected_parachain_state::<R, GI, PI, PID>(&calls[0])?,
-						extract_messages_state::<R, GI, MI, LID>(&calls[1])?,
-					))
-				}
-				return None
+		// Try to check if the tx matches one of types we support.
+		let parse_call = || {
+			let calls = self.expand_call(call)?;
+
+			let mut tx_info = TxData {
+				submit_finality_proof: None,
+				submit_parachain_heads: None,
+				receive_messages_proof: calls
+					.receive_messages_proof
+					.receive_messages_proof_info_for(LID::get())?,
+			};
+			if let Some(submit_parachain_heads) = calls.submit_parachain_heads {
+				tx_info.submit_parachain_heads =
```

Comment:
Not an issue - just sharing my thoughts :) IMO strongly typed `enum CallType` here is better than the struct with 2 `Option`s. With struct you may (accidentally) create it with `submit_finality_proof` set to `Some(_)` and `submit_parachain_heads` set to `None`, which is an issue. Enum forces you to have `submit_parachain_heads` call in the batch if you have the `submit_parachain_heads` call. I had to return to this method twice to ensure that that never happens :) But that's just my option - you're the author, so do what you think is best here.

Reply:
Makes sense. Reverted to using the `enum`.